### DISCUSSION TITLE
MLB

### DISCUSSION THREAD

**Discussion by JP23000** [kyleskom/NBA-Machine-Learning-Sports-Betting#151] [Ideas]
I brought this up in an email with Kyle and he encouraged me to post this here.

How would I go about enticing the members of this group to duplicate this system for MLB before the first month of the season finishes this year?

Thanks!

John

**Comment by nealmick** [comment]
I have considered building out projects for other sports, from what ive seen higher scoring games are a bit more predictable.  I think scores like 110-102 are more specific then 1-2, may effect loss calculation.... Im sure there are ways to overcome this, and might even make it so there are less close calls, and more obvious winners overall.

**Reply by kyleskom** [reply]
This is correct and the reason I choose NBA. NBA is the best to bet on. It has a season with the most games to minimize outliers in data. On top of that it's one of the only sport where top players arnt affected by others. For example an all star wr in the NFL needs a quarterback to get him the ball. A top point guard will always get the ball.

**Reply by JP23000** [reply]
Yes this absolutely makes sense.  This probably also leads to why the NBA has the highest winning rate of ML Vegas favorites and MLB has the lowest.  With a 57.5% ML Favorite win rate on MLB, we would need to get the ML prediction correct ~60% of the time to break even.

I wonder if you put less emphasis on all individual players... Maybe account for each players error rate in fielding... but really look at this as just a pitching vs batting matchup.  How is today's pitcher going to perform against this line-up?  I think when we look at it through this lens it becomes a little easier. Especially if we narrow it down to pitchers who are hot and on a streak.  They become dominating.  It doesn't matter who it is in the batter box; Babe Ruth, Barry Bonds, Sammy Sosa... It doesn't matter.  When Randy Johnson is hot.. He is striking them all out.  So I think being selective in the games that are chosen would be a great way to help increase the overall win rate.

Just thinking out loud.

**Reply by JP23000** [reply]
Again, just thinking out loud about player dependency and their affect on each other....

Do we need to be concerned with players affecting each other?  So a top PG will almost always get the ball, unless he is off the floor or they inbound it to the SG or a team with a Point Center...  But a QB will get every snap he is on the field for.  In order for a player to be at the top of his respective position they need to have more than skills, but stats to back it up.  If they are getting the stats necessary to be identified as a top performer then they should have a system built around him that facilitates that production.  I feel like the effect of the team mate dependency lowers the closer the player gets to the focal point of the game plan.  Am I concerned about WR5 getting yards? Only if I am betting on his player props.  I know DK Metcalf is getting the ball. I can model about how often he will get it as well due to game plans and historical records. This changes with injuries, which there is more of in the NFL than NBA... but Im curious what it would look like if we tried it.  Ill find the link to this medium blog I found where he built a Random Forest NFL ML predictor and it averaged out around 67% correct I think.  Ill post it here shortly.  It could help get the ball rolling.

**Reply by nealmick** [reply]
Good thoughts, as far as model architecture goes, I think first focus on what features are most impactful and most correlated to your target, say just MLB scores are the target for now.  Try to figure out what input features you want in your dataset.... you might think just use all the data, but this could end up with sparse features, and make training difficult... Maybe team stats like win/loss/streak/rank etc....then maybe each teams pitcher stats, and maybe some batter stats, i dont know alot about MLB played little league a long time ago haha.

**Reply by nealmick** [reply]
Good inspiration
https://www.imdb.com/title/tt1210166/

**Reply by JP23000** [reply]
Also something to consider that may be less work.... repurposing this for international basketball as well.  With the Canadian League you get basketball going all the way through the end of July.

**Comment by kyleskom** [comment]
@JP23000 So yes this same idea can be applied to other sports, Hers a quick list of things needed.
1.) Get all team data for every day for all the seasons as well as odds data for those games
2.) You would need to combine that data in the same where as done here and used that data to train the models
3.) Get current data for current days games + odds data
4.) Use that current data to pass to the model to give predictions.

**Reply by nealmick** [reply]
found a dataset on kaggle, going back to 1871, over 50mb of MLB data!  A great start if your interested @JP23000 

https://www.kaggle.com/datasets/saurabhshahane/major-league-baseball-dataset

**Reply by betevents** [reply]
how does the model accounts for double headers?

**Reply by nealmick** [reply]
Good idea!  doubleHeader = False or rest days....

**Reply by kyleskom** [reply]
What exactly is a double header? (I don't follow baseball)

**Reply by betevents** [reply]
basically the same teams play twice in the same day and venue, one a early game and the other one normally 5-6 hours later. the reason I asked about the DH is because the starting pitchers will be different for each game, same with the official 9 batters lineup. but if the model makes a DH a false/rest days then we probably take both games out (false/rest)

**Reply by kyleskom** [reply]
Oh thats surely interesting. Ya we would have to account for that.

**Reply by JP23000** [reply]
Yes, and the problem is a lot of times the games are very unpredictable.  I would suggest skipping all games that are DH.  Weird things happen in those games.

**Reply by JP23000** [reply]
Hi Kyleskom,

So it appears that repurposing this for MLB is way outside the scope of my abilities and experience.  I can pay someone to do this if they wanted to take on this project.

**Reply by nealmick** [reply]
Haha, how much?

**Reply by nortageMliaHllA** [reply]
Hi @kyleskom with the amount of college teams and games played weekly, would applying this ML model to college football require any additional steps? What about NFL?

**Comment by kyleskom** [comment]
If there is enough support for it maybe I can go ahead and add that onto this project. It would take some time though.

**Reply by JP23000** [reply]
So it appears that repurposing this for MLB is way outside the scope of my abilities and experience. I can pay someone to do this if they wanted to take on this project.

**Reply by BitcoinGPT** [reply]
i can help 
-wanted to discuss the feasible plans to [kyleskom] as he a brilliant mind and i believe with trust and transparency we can make Impact of this Beaut'

**Comment by Marcfeitosa** [comment]
I'm in.

I can be of great help in soccer, For soccer I recommend the Poisson distribution model. I would also recommend betting in corners as it is higher than the score itself.

Count me in.
    Em quarta-feira, 15 de março de 2023 às 18:15:44 BRT, Kyle Skompinski ***@***.***> escreveu:  
 
 


If there is enough support for it maybe I can go ahead and add that onto this project. It would take some time though.

**Reply by JP23000** [reply]
This is not the first time I have heard that betting in corners is the best way to go for soccer.

**Reply by betevents** [reply]
bookings too, the ones in south america are really easy to beat both overs and unders.
https://www.soccerbase.com/referees/home.sd

**Reply by BGB2k11** [reply]
Germany 9/10 over 2,5 goals and alot of corners.
Australia : Cards

**Comment by betevents** [comment]
If someone wants to contribute getting stats (I'm busy as hell atm)

Events:
https://statsapi.mlb.com/api/v1/schedule?language=en&sportId=1&startDate={{startDate}}&endDate={{endDate}}
example for today: https://statsapi.mlb.com/api/v1/schedule?language=en&sportId=1&startDate=03/15/2023&endDate=03/15/2023

Loop through [dates]->[games] get the gamePk id

Game (stats)
https://statsapi.mlb.com/api/v1.1/game/{{gamePk}}/feed/live?language=en
Example: https://statsapi.mlb.com/api/v1.1/game/719113/feed/live?language=en    (Rays @ Red Sox)

liveData->boxscore -> teams -> ...

**Comment by Marcfeitosa** [comment]
I'll look for databases in for latin america and europe
    Em quarta-feira, 15 de março de 2023 às 20:33:57 BRT, betevents ***@***.***> escreveu:  
 
 


bookings too, the ones in south america are really easy to beat both overs and unders.
https://www.soccerbase.com/referees/home.sd

**Comment by SamuelLachance** [comment]
https://rdpharr.github.io/project_notes/baseball/benchmark/webscraping/brier/accuracy/calibration/machine%20learning/2020/09/20/baseball_project.html


That project was profitable back in 2020 for me if anyone can update it for the 2023 season you are a legend

**Reply by sportsfan69** [reply]
I've tried to update it for 2023 but it appears that baseball-reference.com keeps blocking my requests after the first few.

**Reply by BGB2k11** [reply]
if you can create a webscraper with a proxy. This can be avoided ;)

**Comment by aaaclones** [comment]
There’s some keywords touched up on which is “profitable in 2020” things
happen fast especially in this field

**Reply by SamuelLachance** [reply]
it was profitable for season 2019-2020 and 2020-2021 but since Baseball reference changed their website the code is broken and no longer work and i'm not good enough to fix it .

**Comment by aaaclones** [comment]
I am sure you can most certainly find the data outside of baseball
reference , if team data is the only values it’s missing it shouldn’t be a
big problem , let me look into it so we can furthermore discuss

**Reply by JP23000** [reply]
Do we have any update on this?  I can maybe try to dig into it to find a new source to pull data from, or we just adjust the existing code to the new baseball reference requirements?

**Reply by mitchellz4** [reply]
Hey all, I am currently working the rdpharr model which is linked above after adjusting the way the html data was parsed I am able to get the data from baseball reference. Can add to a GitHub and share with people for collaboration if you would like?

**Comment by aaaclones** [comment]
I am currently working on something maybe we can colab on this I will tell
you the framework on how it flows

**Comment by aaaclones** [comment]
you can use the existing code for this and transition it to baseball

**Comment by aaaclones** [comment]
sorry i meant you cannot just use the existing code for nba and apply it to
baseball

**Comment by aaaclones** [comment]
Start the repository and share the link

**Reply by mitchellz4** [reply]
I will try to get to this later today, sorry been busy with school and work.

**Comment by aaaclones** [comment]
looking forward to seeing it

**Comment by mitchellz4** [comment]
Yes that has happened with me. Solved this problem by using a VPN.

**Reply by sportsfan69** [reply]
I was using Colab to test out the changes. If you can share what you have when you have time I'd like to try to help out.

**Reply by aaaclones** [reply]
Have you tried pybaseball ?

**Reply by mitchellz4** [reply]
I have shared the repository with you guys let me know if you have any questions or things to suggest. I come from a mechanical engineering background, so my code is not the most organized, I apologize.

**Reply by mike-bostedt** [reply]
Hey guys, I am interested in helping out with this project if I can. I'm a Front End Engineer but I am looking to play with ML in my free time. 

If you're really trying to advance the project quickly should really open source it. Look at how this project has done open source.

**Reply by mitchellz4** [reply]
I am more than welcome to making this open source or open to anyone that wants to help out with the project!
The more the better. I will add you to the repo mike.

**Reply by SamuelLachance** [reply]
@mitchellz4 wow, where did you share it ? data scraping always been a nightmare for me. i worked on a nhl model personnaly and would like to share  , but the model size is 6gb (.pk1)  and idk how to share large file like those on github . I think the max is 2gb by file .

**Reply by mitchellz4** [reply]
sorry everyone, I was taking a few days off from this to gather some new ideas. I will make it public for people to see when I can get it working. But if you would like to see it now or help out I can share just send me a message or reply here!

**Reply by ykk9595** [reply]
I'm very interested can you show me first。
Thank you

**Comment by mitchellz4** [comment]
Will do. I should have some time this evening do it.

**Comment by aaaclones** [comment]
I agree the degree of difficulty will be lowered if you want to make it
open source , It will also speed things up tremendously since a lot of
people will be doing a little ... Doing this project from scratch with were
we are at will take a lot of our time , I am not sure i can invest all my
time on this before we get it to run properly , This is Vegas most
profitable sport dollar for dollar , Nfl and Cfb has the highest volume but
in terms of the sport hardest to beat is baseball ....

**Reply by mike-bostedt** [reply]
I would also like to get this working for American football but with the season a ways off it would be more fun to develop something you can watch. I'm also assuming I can take any learnings from an NBA and MLB project and apply them to NFL.

**Comment by aaaclones** [comment]
We can also take a different approach , Since baseball lineups dont tend to
deviate much outside the starting pitcher we can really simplify it down
into 3 parts .. we can do first 3 innings / 3 innings / and final 3 innings
we can gather all  the data collected on every single player , we can then
divide the game into 3 parts see how the game looks in each of these
segments and then determine some sort of outcome , Basketball is a lot
different since its mainly centered around 6-7 player rotation and it's a
little more linear , but baseball is more singular , each player isn't
effected by one another .... I am not sure how we can go about doing this
but I am most certain theres a way

**Comment by aaaclones** [comment]
essentially it's all machine learning , we can decide what data to use how
to compile and deploy it .. i Just dont have the bandwidth to do it
myself... If we have a group of people that can all chip in that would be
very useful

**Comment by Marcfeitosa** [comment]
I'm already doing something, it is far from being ready but I can share
    Em quarta-feira, 5 de abril de 2023 às 03:23:45 BRT, BGB ***@***.***> escreveu:  
 
 


if you can create a webscraper with a proxy. This can be avoided ;)

**Comment by aaaclones** [comment]
I will help out this weekend

**Comment by Marcfeitosa** [comment]
I can help too
    Em sexta-feira, 7 de abril de 2023 às 08:12:52 BRT, aaaclones ***@***.***> escreveu:  
 
 

I will help out this weekend

**Reply by mitchellz4** [reply]
I just made the repo public for anyone! https://github.com/mitchellz4/MLB-Model

**Comment by aaaclones** [comment]
Do you have a place I can get up to date api keys for nba ?
